File: 0001-two-sum/1.cpp
class Solution {
public:
  vector<int> twoSum(vector<int>& nums, int target) {
    /* Approach:
     *   1. [Bruteforce] -- T.C: O(N^2), S.C: O(1)
     *   2. [Hash Table] -- T.C: O(N), S.C: O(N)
     */

    /* Approach 1 - Bruteforce
    int n = nums.size();
    for (int i = 0; i < n; i++) {
      for (int j = i + 1; j < n; j++) {
        if (target - nums[i] == nums[j])
          return {i, j};
      }
    }
    return {-1, -1};
    */

    // Approach 2 - Hash Table
    int                     n = nums.size();
    unordered_map<int, int> mp;

    for (int i = 0; i < n; i++) {
      auto it = mp.find(target - nums[i]);
      if (it != mp.end())
        return {it->second, i};
      mp[nums[i]] = i;
    }
    return {-1, -1};
  }
};

File: 0001-two-sum/1.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Approach 2: HashMap -- T.C: O(N), S.C: O(N)
        hashmap = {}

        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in hashmap:
                return [i, hashmap[complement]]
            hashmap[nums[i]] = i

        return []

        """
        # Approach 1: Brute Force -- T.C: O(N^2), S.C: O(1)
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if target - nums[i] == nums[j]:
                    return [i, j]
        return []
        """
